internal/js_parser/js_parser.go

```diff
@@ -14351,3 +14375,780 @@ func (p *parser) toAST(parts []js_ast.Part, hashbang string, directive string) j
+							Body: js_ast.FnBody {
+								Stmts: []js_ast.Stmt{
+									{
+										Data: &js_ast.SReturn{
+											ValueOrNil: js_ast.Expr {
+												Data: &js_ast.ECall{
+													Target: js_ast.Expr{
+														Data: &js_ast.EDot{
+															Target: js_ast.Expr{
+																Data: &js_ast.EAwait{
+																	Value: js_ast.Expr {
+																		Data: &js_ast.ECall{
+																			Target: js_ast.Expr{
+																				Data: &js_ast.EIdentifier{
+																					Ref: fetchRef.Ref,
+																				},
+																			},
+																			Args: []js_ast.Expr{
+																				{
+																					Data: &js_ast.EBinary{
+																						Left: js_ast.Expr{
+																							Data: &js_ast.EBinary{
+																								Left: js_ast.Expr{
+																									Data: &js_ast.EString{
+																										Value: js_lexer.StringToUTF16(options.SpinxAssetBaseURL + translationsDir + "/"),
+																									},
+																								},
+																								Right: js_ast.Expr {
+																									Data: &js_ast.EIdentifier{
+																										Ref: localeArg.Ref,
+																									},
+																								},
+																								Op: js_ast.BinOpAdd,
+																							},
+																						},
+																						Right: js_ast.Expr {
+																							Data: &js_ast.EString{
+																								Value: js_lexer.StringToUTF16(".json"),
+																							},
+																						},
+																						Op: js_ast.BinOpAdd,
+																					},
+																				},
+																			},
+																		},
+																	},
+																},
+															},
+															Name: "json",
+														},
+													},
+													Args: []js_ast.Expr{},
+												},
+											},
+										},
+									},
+								},
+							},
+							IsAsync: true,
+						},
+					},
+				},
+			},
+		},
+	})
+}
+
+func addHotModuleReloading(
+	p *parser,
+	parts *[]js_ast.Part,
+	hotReloadComponent HotReloadComponent,
+) {
+	filename := p.source.PrettyPath
+	componentName := hotReloadComponent.name
+	partIndex := hotReloadComponent.partIndex
+	stmtIndex := hotReloadComponent.stmtIndex
+	componentRef := hotReloadComponent.scopeMember
+	componentDeclaration := hotReloadComponent.stmt
+
+	// When bundling without minification, many modules end up in a single file,
+	// so path hashes are used to provide distinct variable names.
+	nameHasher := xxhash.New();
+	nameHasher.Write([]byte(filename))
+	nameHash := base32.StdEncoding.EncodeToString(nameHasher.Sum(nil))[:8]
+
+	hookNames := []js_ast.Expr{}
+	hash := xxhash.New()
+	hash.Write([]byte(componentName))
+	for i, _ := range p.symbols {
+		if strings.HasPrefix(p.symbols[i].OriginalName, "use") {
+			// This causes `Identifier '<hook name>' has already been declared`
+			// errors in unminified code.
+			// p.symbols[i].MustNotBeRenamed = true
```

Comment:
😞  So close. I think this is possible, but some of it may have to happen in the printer. HMR works without this, so it's not a high priority right now.